toolchain/codegen/codegen.cpp

```diff
@@ -0,0 +1,69 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/codegen/codegen.h"
+
+#include <cstdio>
+
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/MC/TargetRegistry.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/TargetSelect.h"
+#include "llvm/Target/TargetMachine.h"
+#include "llvm/Target/TargetOptions.h"
+#include "llvm/TargetParser/Host.h"
+
+namespace Carbon {
+
+void CodeGen::GenerateObjCodeFromModule(llvm::Module& module) {
+  // Initialize the target registry etc.
+  llvm::InitializeAllTargetInfos();
+  llvm::InitializeAllTargets();
+  llvm::InitializeAllTargetMCs();
+  llvm::InitializeAllAsmParsers();
+  llvm::InitializeAllAsmPrinters();
+
+  auto target_triple = llvm::sys::getDefaultTargetTriple();
+
+  std::string error;
+  const auto* target = llvm::TargetRegistry::lookupTarget(target_triple, error);
+
+  if (!target) {
+    output << error;
+    return;
+  }
+
+  const auto* cpu = "generic";
+  const auto* features = "";
+
+  llvm::TargetOptions target_opts;
+  auto reloc_model = std::optional<llvm::Reloc::Model>();
+  auto* target_machine = target->createTargetMachine(
+      target_triple, cpu, features, target_opts, reloc_model);
+  module.setDataLayout(target_machine->createDataLayout());
+  module.setTargetTriple(target_triple);
+
+  // const auto* file_name = "objfile.o";
+  // std::error_code ec;
+  // llvm::raw_fd_ostream dest(file_name, ec, llvm::sys::fs::OF_None);
+  llvm::raw_fd_ostream dest(fileno(stdout), false);
+
+  // if (ec) {
+  //   output << "Could not open file: " << ec.message();
+  //   return;
+  // }
+
+  llvm::legacy::PassManager pass;
```

Comment:
Should this be using a `ModulePassManager` instead of the `legacy::PassManager`?

Reply:
The function takes  `PassManagerBase` object and `ModulePassManager`  doesn't inherit this class.
https://llvm.org/doxygen/classllvm_1_1LLVMTargetMachine.html
https://llvm.org/doxygen/classllvm_1_1legacy_1_1PassManagerBase.html

Reply:
Per discussion, it turns out the legacy pass manager is still used for this detail.
https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/BackendUtil.cpp#L1071-L1074
https://llvm.org/docs/NewPassManager.html#status-of-the-new-and-legacy-pass-managers